File: appli-android/app/src/main/java/com/example/sportapp/core/data/model/FoodPortion.kt
package com.example.sportapp.core.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.sportapp.core.utils.CustomDateUtils.getNowISO8601

/**
 * Nutrition A1 (2026-06-17) — portage Android du modèle serveur `food_portions`
 * (cf. serveur/app/models/food_portion.py). Portions nommées d'un aliment
 * (« 1 œuf = 60 g »). Ownership indirect : FoodPortion -> Food -> User.
 *
 * Pas de `user_id` (le serveur ne l'expose pas, ownership via le Food parent).
 * FK Room vers Food (CASCADE, miroir du serveur).
 */
@Entity(
    tableName = "food_portions",
    indices = [
        Index(value = ["uuid"], unique = true),
        Index(value = ["food_uuid"]),
    ],
    foreignKeys = [
        ForeignKey(
            entity = Food::class,
            parentColumns = ["uuid"],
            childColumns = ["food_uuid"],
            onDelete = ForeignKey.CASCADE
        )
    ]
)
data class FoodPortion(
    @PrimaryKey
    @ColumnInfo(name = "uuid") val uuid: String,

    @ColumnInfo(name = "food_uuid") val foodUUID: String,

    val label: String,                                  // « 1 œuf », « 1 cuillère à soupe »
    val grams: Float,

    @ColumnInfo(defaultValue = "0") val synced: Boolean = false,
    @ColumnInfo(defaultValue = "0") val pendingDeletion: Boolean = false,

    @ColumnInfo(name = "updated_at") val updatedAt: String? = getNowISO8601(),
)

File: serveur/app/models/food_portion.py
# app/models/food_portion.py
#
# Nutrition V1 (2026-06-12, cf. docs/NUTRITION_DESIGN.md §3.2).
# Portions nommees d'un aliment (« 1 oeuf = 60 g »). Ownership indirect :
# FoodPortion -> Food -> User (cascade ownership politique 8).

from sqlalchemy import Column, Integer, String, Float, ForeignKey, DateTime
from app.database import Base
import uuid


class FoodPortion(Base):
    __tablename__ = "food_portions"

    id = Column(Integer, primary_key=True, index=True)
    uuid = Column(String, nullable=False, unique=True, default=lambda: str(uuid.uuid4()))
    food_uuid = Column(String, ForeignKey("foods.uuid", ondelete="CASCADE"), nullable=False, index=True)

    label = Column(String, nullable=False)             # « 1 oeuf », « 1 cuillere a soupe » (user-typed)
    grams = Column(Float, nullable=False)              # equivalent en grammes

    updated_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return (
            f"FoodPortion(id={self.id}, uuid='{self.uuid}', food_uuid='{self.food_uuid}', "
            f"label='{self.label}', grams={self.grams})"
        )
